mesh/mesh_connections.hpp

```diff
@@ -0,0 +1,256 @@
+       @a parents.dim and @a parents.boundary members.  The parent 
+       entity ID numbers are placed in @a parents.ids member.  For the 
+       example mesh with vertex, edge, face/element, and boundary element 
+       IDs we have:
+
+       dim = 0        dim = 1        dim = 2     Boundary Elems 
+      0---1---2      +-0-+-1-+      +---+---+      +-0-+-1-+
+      |   |   |      6   7   8      | 0 | 1 |      4   |   5
+      3---4---5      +-2-+-3-+      +---+---+      +---+---+
+      |   |   |      9  10  11      | 2 | 3 |      6   |   7
+      7---8---9      +-4-+-5-+      +---+---+      +-2-+-3-+ 
+
+                                  children              parents
+                                dim,bdry,ids          dim,bdry,ids
+      ParentsCoveredByEntities({1,false,{0,1,2,6,7}}, {2,false,{}}) ->  {2,false,{0}}, element 1 is not covered
+      ParentsCoveredByEntities({0,false,{0,1,2,3,4}}, {2,false,{}}) ->  {2,false,{0}}
+      ParentsCoveredByEntities({0,false,{0,1,2,3,7}}, {1,true ,{}}) ->  {1,true ,{0,1,4,6}}**/   
+   void ParentsCoveredByEntities(const EntityIDs &children, EntityIDs &parents);
+
+   /** @brief Returns the neighbor entity IDs of the given @a entity.  The @a neighbors 
+       struct operates as an in/out parameter.  The kind of the neighbor entities is set in the 
+       @a neighbors.dim and @a neighbors.boundary members.  The neighbor 
+       entity ID numbers are placed in @a neighbors.ids member.  The @a shared_dim parameter
+       sets the dimension of the entities that the neighbors are sharing across.  Note:  The 
+       entity.dim and neighbors.dim must be the same.  For example if the entity/neighbor dims are 0
+       and the shared_dim is 0 this will return the vertex that are neighbors across the edges connected
+       to the vertex entity.  For the example mesh with vertex, edge, face/element, and boundary element 
+       IDs we have:
+
+       dim = 0        dim = 1        dim = 2     Boundary Elems 
+      0---1---2      +-0-+-1-+      +---+---+      +-0-+-1-+
+      |   |   |      6   7   8      | 0 | 1 |      4   |   5
+      3---4---5      +-2-+-3-+      +---+---+      +---+---+
+      |   |   |      9  10  11      | 2 | 3 |      6   |   7
+      7---8---9      +-4-+-5-+      +---+---+      +-2-+-3-+ 
+
+                            entity                         neighbors
+                         dim,bdry,ids                    dim,bdry,ids
+      NeighborsOfEntity({0,false,1},1,{0,false,{}}) ->  {0,false,{0,2,4}}
+      NeighborsOfEntity({0,false,1},2,{0,false,{}}) ->  {0,false,{0,1,3,4}}
+      NeighborsOfEntity({1,true ,5},0,{1,false,{}}) ->  {0,false,{1,11}},   edge numbering
+      NeighborsOfEntity({1,true ,5},0,{1,true ,{}}) ->  {0,true,{1,7}},     boundary element numbering
+      NeighborsOfEntity({0,false,1},2,{1,false,{}}) ->  error, entity.dim != neighbors.dim**/  
+   void NeighborsOfEntity(const EntityID &entity, int shared_dim, EntityIDs &neighbors);
```

Comment:
Because of the lazy table generation, these can indeed change the object.

Reply:
Just want to leave a design consideration here: I think you could also mark the caches (i.e. tables) mutable, so the method can be marked const, but still mutate the underlying data structure as a side effect. However, I am not sure what would be the better choice.

Reply:
Yeah, I am aware of that pattern and personally loath it.  If I am going to go through the trouble of arranging const correctness, I want the const keyword to 'really' mean it.  If it is only going to kinda mean it I would rather throw const correctness in the dustbin.  I'm open to being convinced that the pattern has value if folks have a good argument for it.

Reply:
I think the motivation for having the internal tables be mutable and the query functions const is something like the following:

* If the tables are not mutable, then `NeighborsOfEntity` cannot be const
* If `NeighborsOfEntity` is not const, then it cannot be called on a non-const `MeshConnections` object
* Presumably the `MeshConnections` object is obtained through an an accessor in the `Mesh` object. A non-const `MeshConnections` cannot be returned from a const Mesh.
* Consequently, none of these topological queries can be made starting from a `const` Mesh. This seems like a limitation, there are many use-cases where one wants to query the mesh topology but not modify (e.g. refine or transform) the mesh.

I think it is better in this case to consider the distinction between "logical" and "physical" state when it comes to const-correctness. Lazily constructing the tables does not modify any externally observable state of the object. There is no change to the mesh topology, geometry, or other "intrinsic" properties. It can be considered simply an implementation detail — we cache (memoize) the connectivities as an optimization.

From the point of view of user code, having a `const Mesh` should mean that the mesh itself can be queried, but not modified, and a `Mesh` object should only be passed to a function if there is some reason to really modify the mesh (do some adaptivity, refinement, etc.). If these internal state changes require using a `Mesh` instead of `const Mesh`, then basically everyone needs to use non-const `Mesh` all the time, and the distinction becomes meaningless.

There are some good explanations in the official C++ FAQ from the standards committee:
* [Should the constness of my public member functions be based on what the method does to the object’s logical state, or physical state?](https://isocpp.org/wiki/faq/const-correctness#logical-vs-physical-const)
* [How can it help me design better classes if I distinguish logical state from physical state?](https://isocpp.org/wiki/faq/const-correctness#logical-vs-physical-state)
* [What do I do if I want a const member function to make an “invisible” change to a data member?](https://isocpp.org/wiki/faq/const-correctness#mutable-data-members)

Reply:
Fair enough, I could see someone wanting to make a const Mesh, and that's good enough for me.